File: .vscode/2 C#/P13/Fuhrparkverwaltung_extended/LinkedList.cs
// LinkedList.cs
using System;

namespace Praktikum13
{
    /// <summary>
    /// Generische einfach verkettete Liste.
    /// Ergänzt um Property Count und Indexer (Aufgabe 1).
    /// </summary>
    /// <typeparam name="T">Typ der zu speichernden Elemente</typeparam>
    public class LinkedList<T>
    {
        #region Innere Klassen

        /// <summary>
        /// Innere Klasse für Listenelemente (Node)
        /// Kapselt die Daten und den Zeiger zum nächsten Element
        /// </summary>
        private class ListNode
        {
            public T Data { get; set; }
            public ListNode? Next { get; set; }

            public ListNode(T data)
            {
                Data = data;
                Next = null;
            }
        }

        /// <summary>
        /// Innere Klasse für Iterator-Implementierung
        /// Implementiert die IIterator Schnittstelle
        /// </summary>
        private class ListIterator : IIterator<T>
        {
            private ListNode current;

            /// <summary>
            /// Konstruktor für Iterator
            /// </summary>
            /// <param name="start">Startknoten für die Iteration</param>
            public ListIterator(ListNode start)
            {
                current = start;
            }

            /// <summary>
            /// Prüft ob weitere Elemente vorhanden sind
            /// </summary>
            /// <returns>true wenn current nicht null ist</returns>
            public bool HasNext()
            {
                return current != null;
            }

            /// <summary>
            /// Liefert das nächste Element und bewegt Iterator weiter
            /// </summary>
            /// <returns>Daten des aktuellen Knotens</returns>
            public T Next()
            {
                if (current == null)
                    throw new InvalidOperationException("Keine weiteren Elemente vorhanden");

                T data = current.Data;
                current = current.Next;
                return data;
            }
        }

        #endregion

        #region Private Felder

        private ListNode? head;    // Erstes Element der Liste
        private int count;        // Anzahl der Elemente

        #endregion

        #region Eigenschaften

        /// <summary>
        /// Eigenschaft für die Anzahl der Elemente in der Liste.
        /// (Aufgabe 1)
        /// </summary>
        public int Count
        {
            get { return count; }
        }

        /// <summary>
        /// (Aufgabe 1) Indexer für den Zugriff auf Elemente der Liste.
        /// </summary>
        /// <param name="index">Der 0-basierte Index des Elements.</param>
        /// <returns>Das Element am angegebenen Index.</returns>
        /// <exception cref="IndexOutOfRangeException">Wird geworfen, wenn der Index außerhalb des gültigen Bereichs liegt.</exception>
        public T this[int index]
        {
            get
            {
                // Prüfe auf gültigen Lesebereich
                if (index < 0 || index >= count)
                {
                    throw new IndexOutOfRangeException($"Index {index} ist außerhalb des gültigen Bereichs [0, {count - 1}].");
                }

                ListNode current = head;
                for (int i = 0; i < index; i++)
                {
                    current = current.Next;
                }
                return current.Data;
            }
            set
            {
                // Die set-Operation des Indexers ist nur für den Index 'Count' gültig, um Elemente am Ende der Liste hinzuzufügen. (Aufgabe 1)
                if (index != count)
                {
                    throw new IndexOutOfRangeException($"Set-Operation ist nur für Index 'Count' ({count}) erlaubt. Ungültiger Index: {index}.");
                }

                // Fügt das Element am Ende hinzu (verhält sich wie Add-Methode)
                Add(value);
            }
        }

        #endregion

        #region Konstruktor

        /// <summary>
        /// Konstruktor, erstellt leere Liste
        /// </summary>
        public LinkedList()
        {
            head = null;
            count = 0;
        }

        #endregion

        #region Öffentliche Methoden

        /// <summary>
        /// Fügt ein Element am Ende der Liste hinzu
        /// </summary>
        /// <param name="item">Hinzuzufügendes Element</param>
        public void Add(T item)
        {
            ListNode newNode = new ListNode(item);

            if (head == null)
            {
                head = newNode;
            }
            else
            {
                // Zum Ende der Liste navigieren
                ListNode current = head;
                while (current.Next != null)
                {
                    current = current.Next;
                }
                // Neuen Knoten am Ende anhängen
                current.Next = newNode;
            }
            count++;
        }

        /// <summary>
        /// Entfernt das Element an der angegebenen Position
        /// </summary>
        /// <param name="index">Index des zu entfernenden Elements (0-basiert)</param>
        /// <returns>true wenn erfolgreich entfernt, false bei ungültigem Index</returns>
        public bool Remove(int index)
        {
            // Prüfung auf gültigen Index
            if (index < 0 || index >= count || head == null)
                return false;

            // Erstes Element entfernen
            if (index == 0)
            {
                head = head.Next;
                count--;
                return true;
            }

            // Element an Position index entfernen
            ListNode current = head;
            // Bis zum Element vor dem zu löschenden navigieren
            for (int i = 0; i < index - 1; i++)
            {
                // Sollte nicht null werden, da Index validiert wurde
                if (current.Next == null) return false;
                current = current.Next;
            }

            // Element überspringen (löschen)
            if (current.Next != null)
            {
                current.Next = current.Next.Next;
                count--;
                return true;
            }

            return false;
        }

        /// <summary>
        /// Erstellt einen Iterator für diese Liste
        /// </summary>
        /// <returns>Neuer Iterator</returns>
        public IIterator<T> GetIterator()
        {
            return new ListIterator(head);
        }

        /// <summary>
        /// Gibt alle Elemente der Liste auf der Konsole aus
        /// Verwendet den Iterator
        /// </summary>
        public void PrintAll()
        {
            if (head == null)
            {
                Console.WriteLine("Liste ist leer");
                return;
            }

            Console.WriteLine("Listeninhalt:");
            IIterator<T> iterator = GetIterator();
            int index = 0;
            while (iterator.HasNext())
            {
                Console.WriteLine($"[{index}]: {iterator.Next()}");
                index++;
            }
        }

        #endregion
    }
}

File: .vscode/1 C++/P6 linkedList/linkedList.cpp
#include "linkedList.h"
#include <cstring>

// Knoten-Struktur für die verkettete Liste
struct Node
{
    char *text; // Zeiger auf den gespeicherten Text
    Node *next; // Zeiger auf den nächsten Knoten

    // Konstruktor für einen neuen Knoten
    Node(const char *t) : next(nullptr)
    {
        if (t)
        {
            text = new char[strlen(t) + 1];
            strcpy(text, t);
        }
        else
        {
            text = nullptr;
        }
    }

    // Destruktor - gibt den Speicher für den Text frei
    ~Node()
    {
        delete[] text;
    }
};

// Konstruktor - erzeugt eine leere Liste
LinkedList::LinkedList() : head(nullptr), tail(nullptr), size(0) {}

// Copy-Konstruktor für die Initialisierung mit existierendem Objekt
LinkedList::LinkedList(const LinkedList &other) : head(nullptr), tail(nullptr), size(0)
{
    Node *current = other.head;
    while (current)
    {
        append(current->text);
        current = current->next;
    }
}

// Zuweisungsoperator
LinkedList &LinkedList::operator=(const LinkedList &other)
{
    if (this != &other)
    {
        // Aktuelle Liste löschen
        clear();

        // Elemente aus other kopieren
        Node *current = other.head;
        while (current)
        {
            append(current->text);
            current = current->next;
        }
    }
    return *this;
}

// Destruktor - gibt allen Speicher frei
LinkedList::~LinkedList()
{
    clear();
}

// Hilfsfunktion zum Löschen aller Elemente
void LinkedList::clear()
{
    while (head)
    {
        Node *temp = head;
        head = head->next;
        delete temp;
    }
    tail = nullptr;
    size = 0;
}

// Fügt den Text am Ende der Liste hinzu (in konstanter Zeit)
int LinkedList::append(const char *text)
{
    if (!text)
        return 0;

    Node *newNode = new Node(text);
    if (!newNode || !newNode->text)
    {
        delete newNode;
        return 0;
    }

    if (!head)
    {
        // Liste ist leer
        head = tail = newNode;
    }
    else
    {
        // Neuen Knoten am Ende anhängen
        tail->next = newNode;
        tail = newNode;
    }

    size++;
    return 1;
}

// Fügt ein Element an Position p ein
int LinkedList::insert(const char *text, int p)
{
    if (!text)
        return 0;

    // Wenn Position >= Listenlänge, am Ende anhängen
    if (p >= size)
    {
        return append(text);
    }

    Node *newNode = new Node(text);
    if (!newNode || !newNode->text)
    {
        delete newNode;
        return 0;
    }

    // Am Anfang einfügen
    if (p <= 0)
    {
        newNode->next = head;
        head = newNode;
        if (!tail)
            tail = newNode;
        size++;
        return 1;
    }

    // An Position p einfügen
    Node *current = head;
    for (int i = 0; i < p - 1 && current; i++)
    {
        current = current->next;
    }

    if (current)
    {
        newNode->next = current->next;
        current->next = newNode;
        if (newNode->next == nullptr)
        {
            tail = newNode;
        }
        size++;
        return 1;
    }

    delete newNode;
    return 0;
}

// Löscht das Element an Position p
int LinkedList::remove(int p)
{
    if (!head)
        return 0;

    // Wenn Position zu groß, letztes Element löschen
    if (p >= size - 1)
    {
        // Letztes Element finden und löschen
        if (size == 1)
        {
            delete head;
            head = tail = nullptr;
        }
        else
        {
            Node *current = head;
            while (current->next != tail)
            {
                current = current->next;
            }
            delete tail;
            tail = current;
            tail->next = nullptr;
        }
        size--;
        return 1;
    }

    // Erstes Element löschen
    if (p <= 0)
    {
        Node *temp = head;
        head = head->next;
        if (!head)
            tail = nullptr;
        delete temp;
        size--;
        return 1;
    }

    // Element an Position p löschen
    Node *current = head;
    for (int i = 0; i < p - 1 && current->next; i++)
    {
        current = current->next;
    }

    if (current->next)
    {
        Node *temp = current->next;
        current->next = temp->next;
        if (temp == tail)
        {
            tail = current;
        }
        delete temp;
        size--;
        return 1;
    }

    return 0;
}

// Liefert den Text an Position p
const char *LinkedList::get(int p)
{
    if (p < 0 || p >= size || !head)
        return nullptr;

    Node *current = head;
    for (int i = 0; i < p && current; i++)
    {
        current = current->next;
    }

    return current ? current->text : nullptr;
}

// Berechnet den Index des ersten Elements mit dem gesuchten Text
int LinkedList::index_of(const char *text)
{
    if (!text || !head)
        return -1;

    Node *current = head;
    int index = 0;

    while (current)
    {
        if (strcmp(current->text, text) == 0)
        {
            return index;
        }
        current = current->next;
        index++;
    }

    return -1;
}

// Liefert den Text des ersten Elements (in konstanter Zeit)
const char *LinkedList::first()
{
    return head ? head->text : nullptr;
}

// Liefert den Text des letzten Elements (in konstanter Zeit)
const char *LinkedList::last()
{
    return tail ? tail->text : nullptr;
}

// Besucht alle Elemente und ruft für jedes die Funktion work auf
void LinkedList::visit_all(void (*work)(const char *t))
{
    if (!work)
        return;

    Node *current = head;
    while (current)
    {
        work(current->text);
        current = current->next;
    }
}

// Hilfsfunktion für Tests - gibt die Größe zurück
int LinkedList::getSize() const
{
    return size;
}